Schema 2.0

edwardchapin opened the pull request:
This is work that was done by Allan Brighton. I'd like to get it merged into master so that we switch over to the new schema as soon as possible.

This will affect how our tech notes look as the python script (parsemodelfiles) in the NIC repo won't be extracting all of the information it needs. However, it does not seem to break the builds, so I think it's reasonable to merge this branch first, and then fix the document generation after.

Commit:
Changes for icd-0.14, JSON schema version 2.0 (Please review alarms, since new fields were added)

Commit:
Changes for icd-0.17, Removed minRate and archiveRate: Only use maxRate now

Commit:
Merge branch 'master' into schema-2.0

# Conflicts:
#	csro/csro-env-assembly/component-model.conf
#	csro/oiwfs-adc-assembly/publish-model.conf
#	csro/oiwfs-detector-assembly/publish-model.conf
#	csro/oiwfs-poa-assembly/publish-model.conf

Commit:
Updated for icd-1.1.0 changes: "-" no longer allowed in component names, prefix removed from component-model.conf

abrighton committed:
Merged changes from master and updated for schema version 2.0

abrighton committed:
Merge branch 'master' into schema-2.0

# Conflicts:
#	csro/oiwfs-env-assembly/command-model.conf
#	csro/oiwfs-env-assembly/publish-model.conf
#	drs/drs-assembly/subscribe-model.conf
#	imager/adc-assembly/publish-model.conf
#	imager/odgw-assembly/command-model.conf
#	imager/odgw-assembly/component-model.conf
#	imager/odgw-assembly/publish-model.conf
#	imager/odgw-assembly/subscribe-model.conf

abrighton committed:
Alarm field name change: acknowledge -> autoAck (reverse meaning)

abrighton committed:
Merging changes from master

abrighton committed:
Merge branch 'master' into schema-2.0

# Conflicts:
#	csro/oiwfs-detector-assembly/command-model.conf
#	csro/oiwfs-detector-assembly/publish-model.conf

abrighton committed:
Fixed duplicate event blocks

zonca commented on csro/oiwfs-adc-assembly/component-model.conf:
```diff
@@ -1,8 +1,8 @@
 
-prefix = iris.oiwfs.adc
-modelVersion = "1.0"
+//prefix = iris.oiwfs.adc
```
is `//` commenting lines out? what is the reason?

edwardchapin replied:
I think @abrighton was drawing our attention to the fact that prefix was going away. You're right though, we should get rid of these comments as part of this PR.

I'll sort out the merge conflicts first, then do another stand-alone commit removing comments.

edwardchapin replied:
OK, one of my recent commits cleans up both these commented-out prefix lines, and updates the component name following Takashi's suggestion.

zonca commented on csro/oiwfs-adc-assembly/publish-model.conf:
```diff
@@ -237,7 +233,6 @@ publish {
 
       The ADC will shift an image position in a manner that depends upon the current state of the ADC. The estimated image shift is published by IRIS for the TCS so that the TCS can adjust the IRIS OIWFS position demands accordingly. The reference wavelength used to calculate the shift for each of the WFS is tcs.cm.iris.oiwfsXAtmDispersion.referenceWavelength.
       """
-      minRate = 1
```
why are the `minRate` values being removed? is it worth writing comments in the files to explain changes?

edwardchapin replied:
Good question. I went and looked... minRate was explicitly removed from the schema last December. I guess it wasn't useful to them: https://github.com/tmtsoftware/icd/commit/b17df023cd9cd7c37ccc685266ae114076cd5a08

zonca replied:
ok, it would be useful to have this information included in the commit message about this change so that it becomes traceable.

edwardchapin replied:
I guess I could add it as a general comment (along with the other changes described in this PR) in the final merge commit of this branch?

zonca commented on csro/oiwfs-adc-assembly/publish-model.conf:
```diff
@@ -309,8 +316,14 @@ publish {
                     </ul>)"""
-      severity = major
-      archive = true
+      // TODO FIXME: The fields below were automatically added and should be updated manually
```
updated manually by whom? is it worth making those updates directly in this PR before merging?

edwardchapin replied:
I think you're right, we should review all of these as part of the PR (and then remove the comments). I'll take a look at this.

edwardchapin replied:
OK, I reviewed and fixed up the TODO comments in alarms for the CSRO assemblies. However, I think @tnakamot and @jimansohn will have to deal with their own assemblies as I don't know them well enough to guess at how the alarms should be formatted.

zonca commented on csro/oiwfs-env-assembly/command-model.conf:
```diff
@@ -232,13 +232,13 @@ At Completion:
+        description      = """Instructs the OIWFS enclosure to open or close the gate valve.
 
-This will tell the enclosure to change the state of the gate valve at the bottom of the snout.  The valve can either be set to be opened or closed and the result is dependent on the current conditions inside and outside of the enclosure.  If conditions will pose a hazard to the enclosure conditions (such as the enclosure is running cold and the OIWFS is not connected to NFIRAOS, then the command  ...
+This will tell the enclosure to change the state of the gate valve at the bottom of the snout.  The valve can either be set to be opened or closed and the result is dependent on the current conditions inside and outside of the enclosure.  If conditions will pose a hazard to the enclosure conditions (such as the enclosure is running cold and the OIWFS is not connected to NFIRAOS, then the command  ...
```
what about splitting this long line? it must be a pain in an editor

edwardchapin replied:
I tend to use editors that wrap the line when displayed (vscode in this case). Not to say that you make a bad suggestion, but at the moment I don't want to go down the road of enforcing a particular formatting style on these model files.

zonca replied:
programmers discuss 80 vs 120 characters, here we have a line of 875 characters. ;)

zonca reviewed (commented):
I am not much familiar with the format, some of my comments could be naive. sorry about that.

Also, there are conflicts, this branch needs a rebase or a merge with master.

tnakamot commented:
I agree that we can fix the document generation script later (I actually have some fixes, not necessarily related to this topic).

One thing I would like to mention is that simply removing "prefix" and replacing "-" with "_" in the component name is not a good idea semantically. [The ICDDB manual](https://docushare.tmt.org/docushare/dsweb/Get/Version-111751/OSW%20TN018-ICDDatabaseUserManual_REL04.pdf) says that the component field is 

> The name of the component. Will be displayed in documents. The component name may contain spaces and dots, but, as of
modelVersion 2.0, not a dash “-“. The prefix for a component, which is needed to subscribe to events, is defined as subsystem.component.

If I interpret correctly, the component name should be defined in a way that the "subsystem.component" becomes the prefix for the event service. So, what we actually need is 
- remove the old component name
- replace field name "prefix" with "component" and strip "iris." at the beginning.

For example, if the component-model.conf has the following two lines,

> component = oiwfs-adc-assemby
> prefix = iris.oiwfs.adc

then, it should be converted to this single line.

> component = oiwfs.adc

Then, the prefix we will use in the Event Service will be "iris.oiwfs.adc", which is originally what we meant when we made all our model files.

I only considered the definition of the field, and this change may have some impact on our documents (or maybe not). Fortunately, we have not developed any software yet, so the impact should be small at this moment. The benefit of this way of defining the component name is that the prefix required for the event service will look like "iris.oiwfs.adc". It looks more natural than "iris.oiwfs_adc_assembly".

At least, this is how I fixed manually in my local edit for just ODGW and Imager detectors. I don't mind if I need to follow this Allan's way to change, but I think we need to agree on it.

edwardchapin commented:
@tnakamot : I agree with your component / prefix suggestions (I never liked having the subsystem included in the component name). I have some issues to look at that were raised by @zonca first, and then I'll tackle the component name updates in a separate stand-alone commit.

edwardchapin committed:
Merge branch 'master' into schema-2.0.

I originally attempted this as a rebase on to master,
but it got too messy given the age of the schema-2.0
branch, so in the end I did it as a merge.

edwardchapin committed:
oiwfsDetector: fix up some alarms for schema-2.0

edwardchapin committed:
Update component names to new dot notation

Previously we had things like imager_odgw_assembly. In the new
schema the prefix is now automatically generated as
subsystem.component, so we have decided that the component name
should simply be the suffix of the prefix we had previously been
using once the leading subsystem name has been stripped.

This commit breaks relationships like subscribed values and
commands. Those will be fixed in later commits.

jimansohn commented:
Regarding the component names, I agree with Takashi's commend and I also suggest using the period(.) instead of the underscore(_). According to the ICDDB manual which Takashi linked, it seems like the prefix will be autogenerated using period, and I personally prefer to be concise with it. 
Also, I want to suggest an idea of removing the word "assembly" from the component name. I think software assembly being the main component in IRIS software it's not crucial to put the word "assembly" in the component name. The type is identified under the "componentType" attribute, anyway.
Once decided/agreed, I'll make the changes for my assemblies after the merge.

edwardchapin commented:
@jimansohn - yes, I'm going to drop HCD / Assembly from the component names.

I'm actually making all the changes myself... almost done. I'll push and ask you to take a look before merging the PR.

edwardchapin committed:
Update component names ub command/subscribe/publish files, and fix inter-assembly interfaces

edwardchapin committed:
csro: remove TODO comments in regards to alarms
